lib/ui/painting/image_decoder_impeller.cc

```diff
@@ -323,59 +298,104 @@ static std::pair<sk_sp<DlImage>, std::string> UnsafeUploadTextureToPrivate(
   blit_pass->SetLabel("Mipmap Blit Pass");
   blit_pass->AddCopy(impeller::DeviceBuffer::AsBufferView(buffer),
                      dest_texture);
-  if (texture_descriptor.size.MipCount() > 1) {
+  if (texture_descriptor.mip_count > 1) {
     blit_pass->GenerateMipmap(dest_texture);
```

Comment:
This is unnecessary and wasteful. The mip chain needs to be generated for the resize texture after the base level from the dest texture has been used as the blit source. Generating a mip chain for the dest texture does nothing.

Reply:
I might be misunderstanding how this works. I think what I actually need to do is blit starting from the closest larger size mip level, otherwise I'm going to drop rows of data, right?

Reply:
Hmm, I may be misunderstanding this now. Can you not use the blit pass to resize? From the docs: "The textures can be different sizes as long as the larger texture has a mipmap level that’s the same size as the smaller texture’s level 0 mipmap.".

In that case, the resize must happen via a draw call.

Reply:
And then you can use a any sampling mode you want.

So, draw from the source to the dest and then generate the mips just once.

Reply:
https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/vkCmdBlitImage.html says

> As the sizes of the source and destination extents can differ in any dimension, texels in the source extent are scaled and filtered to the destination extent

I'm trying to avoid using a draw if possible...

Reply:
> These coordinates are used to sample from the source image, as described in [Image Operations chapter](https://registry.khronos.org/vulkan/specs/1.3-extensions/html/vkspec.html#textures), with the filter mode equal to that of filter, a mipmap mode of VK_SAMPLER_MIPMAP_MODE_NEAREST and an address mode of VK_SAMPLER_ADDRESS_MODE_CLAMP_TO_EDGE. Implementations must clamp at the edge of the source image, and may additionally clamp to the edge of the source region.

Reply:
I just need to specify a filter in blit_pass_vk. But since I only copy mip level 0 I still need to regenerate the mip chain.

Reply:
Oh but we use vkcopyImage hmmm

Reply:
I can detect that the sizes don't match and do the full blit. Unknown if https://developer.apple.com/documentation/metal/mtlblitcommandencoder/1400754-copyfromtexture supports resizing but I could use a MPS instead.

Reply:
Spoiler: it does not, switched to use MPS